### PR TITLE
TASK: Workspace aware commands events migration

### PR DESCRIPTION
**Pull request by mhsdesign**
See https://github.com/neos/neos-development-collection/pull/4708

### CODE REVIEW

File: Neos.ContentRepositoryRegistry/Classes/Command/MigrateEventsCommandController.php
Comment (ahaeslich) [inline Neos.ContentRepositoryRegistry/Classes/Command/MigrateEventsCommandController.php]:
```suggestion
     * Included in March 2024 - before final Neos 9.0 release
```

**Review by nezaniel — APPROVED** [review]
Nicely done

**Review by ahaeslich — CHANGES_REQUESTED** [review]
Tested it locally and run into this error while replaying the contentStream projection:

```
Undefined array key "workspaceName" in /application/Packages/Neos/Neos.ContentRepository.Core/Classes/Feature/WorkspacePublication/Dto/NodeIdToPublishOrDiscard.php line 46
```

**Review by ahaeslich — APPROVED** [review]

**Comment by mhsdesign** [comment]
@ahaeslich looks perfect thanks, we will merge this tomorrow once we have @bwaidelich conceptional approval.

File: Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]: nitpick: We have the guideline to use `find*` for methods that can return `null`
```suggestion
    public static function getContentStreamIdForWorkspace(ContentRepository $contentRepository, WorkspaceName $workspaceName): ContentStreamId
```

File: Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]: "withContentStreamToUse" is a bit bumpy.. I can't think of something better atm.
But more importantly we should add a doc comment explaining this unusual construct even though it is just used internally

File: Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]:
```suggestion
        } finally {
            self::$contentStreamIdToUse = null;
        }
```

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php
Comment (bwaidelich) [inline Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php]: another nitpicky suggestion:
```suggestion
            . '_bkp_' . date('Y_m_d_H_i_s');
```

**Review by bwaidelich — COMMENTED** [review]
Looks great already, just some (partly quite nitpicky) suggestions

**Comment by ahaeslich** [comment]
For documenation: replaying all projects now worked on the projects database.

File: Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]: this line can now be removed, sorry forgot that

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php
Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php]: followed the pattern of another migration, please adjust also there if here ^^ // cc @ahaeslich

File: Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]: I would actually avoid `with` as it doesn't follow our usual pattern of how `with` work (aka return a new instance with the property applied).
how about `ContentStreamIdOverride::applyContentStreamId` or even `applyContentStreamIdToClosure` to be very specific.

Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]: yeah sounds fine the `with` pattern is actually overloaded if you think of the `withoutAuthentication` function which also accepts a closure.

I sadly havent found yet a good prefix for these kinds of functions that set state and cleanup while another function is called.

**Review by ahaeslich — APPROVED** [review]

**Pull request merged**
